src/traits.rs

```diff
@@ -2,43 +2,282 @@
+///         &mut self,
+///         seed: &test_case::Interesting
+///     ) -> Result<Box<Any + Send>> {
+/// #       let count_number_of_lines_in_file = |_| -> Result<_> {  unimplemented!() };
+///         let num_lines = count_number_of_lines_in_file(seed)?;
+///         Ok(Box::new((0, num_lines)))
+///     }
+///
+///     fn clone_state(&self, state: &Box<Any + Send>) -> Box<Any + Send> {
+///         Box::new(Self::downcast(state))
+///     }
+///
+///     // Advancing the state is done by incrementing the index of the next
+///     // line to try removing.
+///     fn next_state(
+///         &mut self,
+///         seed: &test_case::Interesting,
+///         prev_state: &Box<Any + Send>
+///     ) -> Result<Option<Box<Any + Send>>> {
+///         let (index, num_lines) = Self::downcast(prev_state);
+///         Ok(Some(Box::new((index + 1, num_lines))))
+///     }
+///
+///     // When we know that the previous state resulted in an interesting test
+///     // case, we don't need to recount the number of lines in the file or start
+///     // removing lines at index zero again. We know there is only one fewer line
+///     // in the new seed, and that we already tried removing the `index - 1`
+///     // lines from the old seed.
+///     fn next_state_on_interesting(
+///         &mut self,
+///         _new_seed: &test_case::Interesting,
+///         _old_seed: &test_case::Interesting,
+///         prev_state: &Box<Any + Send>
+///     ) -> Result<Option<Box<Any + Send>>> {
+///         let (index, num_lines) = Self::downcast(prev_state);
+///         Ok(Some(Box::new((index - 1, num_lines - 1))))
+///     }
+///
+///     // Remove the line at the given state's current index from the seed
+///     // file. If the index is greater than or equal to the number of lines in
+///     // the seed, then we're out of lines to try removing.
+///     fn reduce(
+///         &mut self,
+///         seed: &test_case::Interesting,
+///         state: &Box<Any + Send>
+///     ) -> Result<Option<test_case::PotentialReduction>> {
+///         let (index, num_lines) = Self::downcast(state);
+///
+///         if index >= num_lines {
+///             return Ok(None);
+///         }
+///
+/// #       let remove_nth_line = |_, _| -> Result<_> { unimplemented!() };
+///         let reduction = remove_nth_line(index, seed)?;
+///         Ok(Some(reduction))
+///     }
+///
+///     // If `preduce` would like to skip further ahead, we can do it more
+///     // efficiently than calling `next_state` in a loop, so override the
+///     // provided default implementation of `fast_forward_states`.
+///     fn fast_forward_states(
+///         &mut self,
+///         _seed: &test_case::Interesting,
+///         n: usize,
+///         prev_state: &Box<Any + Send>
+///     ) -> Result<Option<Box<Any + Send>>> {
+///         let (index, num_lines) = Self::downcast(prev_state);
+///         Ok(Some(Box::new((index + n, num_lines))))
+///     }
+/// }
+/// ```
```

Comment:
cc @mateon1: What do you think of this? How happy would you be to see these methods essentially translated into JSON request/response for the IPC protocol?

Reply:
I like this idea, but I don't think JSON is a good choice for the protocol, and it includes significant code overhead. I believe a CSV-like protocol (with spaces, tabs, commas, or something else as separators) is a good choice.

How will a shell script negotiate what state it stores? Will it just be str?
Is `fast_forward_states` the thing called with DRY? How would ignoring DRY look like? Implementing fast_forward as an identity function?
How will `next_state_on_interesting` look like in IPC? It needs to pass at least the new seed/new interesting testcase, either as a file on disk, or as a sequence number/identifier. Alternatively as a state that generated that file.

For reference, is this how I might implement a chunking lines reducer that (for some reason) ignores DRY?:
```rust
State = (ChunkSize, Position);
fn fast_forward(_, _, state) { state }
fn new_seed(seed, state) { (seed.lines.len(), 0) }
fn next_state(seed, state) {
  let pos = state.size + state.pos;
  if pos > seed.lines.len() { return (state.size / 2, 0); }
  return (state.size, pos)
}
fn interesting_next(old_seed, interesting_state) {
  let pos = interesting_state.pos;
  let new_len = max(old_seed.lines.len() - interesting_state.size, pos);
  if pos == new_len { return (min(new_len, old_seed.0 / 2), 0); }
  return interesting_state;
}
fn reduce(...) {...}
```

Reply:
> I like this idea, but I don't think JSON is a good choice for the protocol, and it includes significant code overhead. I believe a CSV-like protocol (with spaces, tabs, commas, or something else as separators) is a good choice.

I want writing reducer scripts to be *very* easy.

Protocols that are so simple that it is easier to just write an implementation of it (like the existing line-based protocol) than to use a library fit this bill.

JSON protocols fit this bill because every language has a high quality JSON parser, and in a certain sense, JSON is pretty much a lingua franca exchange format. People are familiar with it and don't have to spend time understanding what is going on.

A CSV/TSV/etc protocol are starting to push the limits of what I'd call "very easy". It is less familiar and that will bounce people away immediately. There won't be as many off-the-shelf libraries for dealing with it.

As we've discussed before, extending the line-based protocol is out of the question too, because we need the new protocol to do more things than the old protocol did, and I don't think something like the existing line-based protocol would still be dead-easy to implement either.

> How will a shell script negotiate what state it stores? Will it just be str?

I was imagining any arbitrary JSON value.

> Is fast_forward_states the thing called with DRY? How would ignoring DRY look like? Implementing fast_forward as an identity function?

Yes, exactly.

> How will next_state_on_interesting look like in IPC?

The request from `preduce` would look something like this:

```json
{
    "type": "next_state_on_interesting",
    "old_seed": "/tmp/path/to/old/seed",
    "new_seed": "/tmp/path/to/new/seed",
    "state": <arbitrary JSON>,
}
```

and then the script's response would look something like this:

```json
{
    "state:" <arbitrary JSON>.
}
```

for the `Some` case, or

```json
null
```

for the `None` case.

> For reference, is this how I might implement a chunking lines reducer that (for some reason) ignores DRY?:

Yeah, that looks pretty much right. `next_state` needs a termination case, but this is just a sketch ;)

Reply:
In that case, I believe this is fine. I suggest that the JSON objects be compacted onto a single line, so implementation turns into `JSON.parse(stdin.readline())`, as otherwise it will require tricky implementation to count {} braces (and ignore them in string literals!)

Reply:
> I suggest that the JSON objects be compacted onto a single line

👍